### PR TITLE
New seq2seq lab to replace Tensor2Tensor Poetry lab

### PR DESCRIPTION
**Pull request by vijaykyr**
Note model_deploy.ipynb depends on AI Platform support for 2.0, so will need to be re-reviewed once that is released.

### CODE REVIEW

File: courses/machine_learning/deepdive/09_sequence_keras/translate/translate_model.ipynb [hunk 1]
```diff
@@ -0,0 +1,809 @@
+{
```
Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/translate_model.ipynb]: <p>I'd sort the imports alphabetically within each group.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/translate_model.ipynb]: <p> SEED=0 --&gt; SEED = 0 (pep8) and similarly for LOAD_CHECKPOINT</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/translate_model.ipynb]: <ul><li>formatting (pep 8):</li></ul><p><br></p><p>URL = 'http://storage.googleapis.com/download.tensorflow.org/data/spa-eng.zip'</p><p><br></p><p>path_to_zip = tf.keras.utils.get_file(</p><p>&nbsp;&nbsp;'spa-eng.zip',</p><p>&nbsp;&nbsp;origin=URL,</p><p>&nbsp;&nbsp;extract=True</p><p>&nbsp;&nbsp;)</p><p><br></p><p>path_to_file = os.path.dirname(path_to_zip) + "/spa-eng/spa.txt"</p><p><br></p><ul><li>also: os.path.join would be better than "path1" + "path2" arithmetic, avoiding the problems with "/" </li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/translate_model.ipynb]: <ul><li>I'd add in point 4.: "at the beginning and at the end of each sentence" (if that's what the function does, and where it adds &lt;start&gt; and &lt;end&gt; if not)</li><li>"subsequent cells" -&gt; "subsequent cells:"</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/translate_model.ipynb]: <p>It would be nicer (i.e., simpler to use and less error-prone) to the client code if the ".encode('utf-8')" was taken care of in the body of "utils_preproc.preprocess_sentence".</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/translate_model.ipynb]: <ul><li>The list comprehension needs to be broken up (pep 8).</li><li>Why not using the built-in "open" here? Also, don't you need to close the stream when you are done reading from it (or use it in a context: "with etc.")?</li><li>maybe use .readlines() instead of read/strip/split</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/translate_model.ipynb]: <p>great names! I'd go all the way:</p><ul><li>targ_lang -&gt; target_language</li><li>inp_lang -&gt; input_language</li><li>inp_lang_tokenizer -&gt; input_language_tokenizer</li><li>etc.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/translate_model.ipynb]: <p>I'd remove the comments from the code since your code is already self-documented here. I'd put a paragraph in markdown if you think something more needs to be explained. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/translate_model.ipynb]: <p>Same comment as above + pep 8 (lines are too long and need to be broken up)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/translate_model.ipynb]: <p>I'd put at least a blank line between the blocks of prints to more it easier on the eyes.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/translate_model.ipynb]: <ul><li>I'd run this cell through a pep8 formatter.</li><li>In a markdown cell above the definition of the function, I think it would be useful to describe briefly what we are doing in this function (the general strategy) and explain briefly what tf.roll does. </li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/translate_model.ipynb]: <ul><li>pep8</li><li>I'd put the hyperparameters/constants (e.g., embedding_dim, untits, etc.) defined at the beginning of the cell in upper-case so that they are immediately recognizable from the rest of the code</li><li>Why not packaging the encoder,  decoder, and model code, into their own functions, say, create_encoder(...), create_decoder(...), create_model(...) (or classes possibly)? Is there a reason not to do that here?</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/translate_model.ipynb]: <ul><li>pep 8</li><li>I don't find this cell very readable, but I don't  see right now how to improve on what you did here. Do you see a way to break/package the code in smaller chunks? Maybe separating the code that creates the individual layers from the code that composes these layers? The same kind of comments goes for the cell above. What do you think?</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/488/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

File: courses/machine_learning/deepdive/09_sequence_keras/translate/utils_preproc.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        return ''.join(c for c in unicodedata.normalize('NFD', s)
+            if unicodedata.category(c) != 'Mn')
+
+def preprocess_sentence(w):
+    w = unicode_to_ascii(w.lower().strip())
+
+    # creating a space between a word and the punctuation following it
+    # eg: "he is a boy." => "he is a boy ."
+
+    w = re.sub(r"([?.!,¿])", r" \1 ", w)
```
Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/utils_preproc.py]: * It seems that the documentation is not in sync with the regex. Shouldn't it be rather:  'Surrounding punctuations with a single space: "he is a boy." => "he is a boy . " ' If it's not he behavior you want, then the regex maybe more something like: r" \1" (i.e. no space after group \1)
* Why do we want to treat other punctuation signs differently like ";:,"? Have they been stripped already?

File: courses/machine_learning/deepdive/09_sequence_keras/translate/utils_preproc.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    w = re.sub(r'[" "]+', " ", w)
+
+    # replacing everything with space except (a-z, A-Z, ".", "?", "!", ",")
+    w = re.sub(r"[^a-zA-Z?.!,¿]+", " ", w)
```
Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/utils_preproc.py]: * Shouldn't that line go before w = re.sub(r'[" "]+', " ", w)? My reasoning is that you first replace whatever is not a char or a final punctuation sign by a space, and then you remove the extra spaces created. Not the reverse.

File: courses/machine_learning/deepdive/09_sequence_keras/translate/utils_preproc.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    w = w.rstrip().strip()
+
+    # adding a start and an end token to the sentence
+    # so that the model know when to start and stop predicting.
+    w = '<start> ' + w + ' <end>'
+    return w
+
+def tokenize(lang,lang_tokenizer=None):
```
Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/utils_preproc.py]: * Add a space after the comma
* "lang" is confusing to me, since it is an abbreviation for language. What about "sentences" instead?
* Why "lang_tokenizer" and not simply "tokenizer"?

File: courses/machine_learning/deepdive/09_sequence_keras/translate/utils_preproc.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    
+    Returns:
+    tensor -- int tensor of shape (NUM_EXAMPLES,MAX_SENTENCE_LENGTH)
+    lang_tokenizer -- keras_preprocessing.text.Tokenizer
+    """
+    sentences = [preprocess_sentence(sentence) for sentence in sentences]
+    tokens, _ = tokenize(sentences,tokenizer)
+    return tokens
+
+def int2word(tokenizer,int_sequence):
```
Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/utils_preproc.py]: * Add a space after the comma

File: courses/machine_learning/deepdive/09_sequence_keras/translate/utils_preproc.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    """Converts integer representation to natural language representation
+    
+    Arguments:
+    tokenizer -- keras_preprocessing.text.Tokenizer
+    int_sequence -- an iterable or rank 1 tensor of integers
+    
+    Returns list of string tokens
+    """
+    return [tokenizer.index_word[t] if t!=0 else '' for t in int_sequence]
```
Comment (BenoitDherin) [inline courses/machine_learning/deepdive/09_sequence_keras/translate/utils_preproc.py]: * Add a space before and after "!="
